pkg/imgutil/converter/zstd.go

```diff
@@ -40,34 +41,35 @@ func ZstdLayerConvertFunc(options types.ImageConvertOptions) (converter.ConvertF
 			// No conversion. No need to return an error here.
 			return nil, nil
 		}
-		uncompressedDesc := &desc
-		if !uncompress.IsUncompressedType(desc.MediaType) {
-			var err error
-			uncompressedDesc, err = uncompress.LayerConvertFunc(ctx, cs, desc)
-			if err != nil {
-				return nil, err
-			}
-			if uncompressedDesc == nil {
-				return nil, fmt.Errorf("unexpectedly got the same blob after compression (%s, %q)", desc.Digest, desc.MediaType)
-			}
-			defer func() {
-				if err := cs.Delete(ctx, uncompressedDesc.Digest); err != nil {
-					log.L.WithError(err).WithField("uncompressedDesc", uncompressedDesc).Warn("failed to remove tmp uncompressed layer")
-				}
-			}()
-			log.L.Debugf("zstd: uncompressed %s into %s", desc.Digest, uncompressedDesc.Digest)
+		var err error
+		// Read it
+		readerAt, err := cs.ReaderAt(ctx, desc)
+		if err != nil {
+			return nil, err
 		}
+		defer readerAt.Close()
+		sectionReader := io.NewSectionReader(readerAt, 0, desc.Size)
 
 		info, err := cs.Info(ctx, desc.Digest)
 		if err != nil {
 			return nil, err
 		}
-		readerAt, err := cs.ReaderAt(ctx, *uncompressedDesc)
-		if err != nil {
-			return nil, err
+		labelsMap := info.Labels
+
+		var oldReader io.Reader
+		// If it is compressed, get a decompressed stream
+		if !uncompress.IsUncompressedType(desc.MediaType) {
+			decompStream, err := compression.DecompressStream(sectionReader)
+			if err != nil {
+				return nil, err
+			}
+			defer decompStream.Close()
+			oldReader = decompStream
+			delete(labelsMap, labels.LabelUncompressed)
```

Comment:
Note that I am not quite sure I understand why containerd is doing that in the helper - but I wanted to make sure we have iso behavior.

Reply:
Please do not delete this but update this value if the actual uncompressed digest differs from the original one. 

This is defined at: https://github.com/containerd/containerd/blob/4a18adcfca4e37e829c0d3174bd04d4cbf222f9a/pkg/labels/labels.go#L21
This is used several use cases for example diff computation: https://github.com/containerd/containerd/blob/4a18adcfca4e37e829c0d3174bd04d4cbf222f9a/core/images/diffid.go#L47 
Other clients also might use it.

Reply:
Ok.

Curious what your take is on the following in that context - as I said, not sure what there is to gain in doing it containerd?
https://github.com/containerd/containerd/blob/v1.7.18/images/converter/uncompress/uncompress.go#L80

Anyhow, re-pushed.